src/RequestProcessor.cs

```diff
@@ -281,6 +337,151 @@ private void BindOutputFromResult(InvocationResponse response, AzFunctionInfo fu
+                    var errorMsg = string.Format(PowerShellWorkerStrings.InvalidModuleMajorVersion, moduleName, majorVersion);
+                    var argException = new ArgumentException(errorMsg);
+                    logger.Log(LogLevel.Error, errorMsg, argException, isUserLog: true);
+
+                    throw argException;
+                }
+
+                // Before installing the module, check to see if it is already installed at the given path.
+                var moduleVersionFolderPath = Path.Combine(path, moduleName, latestVersion);
+                if (Directory.Exists(moduleVersionFolderPath))
+                {
+                    // The latest version is already installed, log and return.
+                    var logMsg = string.Format(PowerShellWorkerStrings.ModuleIsAlreadyInstalled, moduleName, latestVersion);
+                    logger.Log(LogLevel.Trace, logMsg, isUserLog: true);
+
+                    return;
+                }
+
+                // Save-Module is able to download versions of Az module side-by-size. However, the dependent modules get
+                // overwritten with the latest version. As a result, the previous Az version will not longer work.
+                // If this is the case, empty the Az directory before running Save-Module.
+                if (moduleName.Equals(AzModuleName, StringComparison.OrdinalIgnoreCase))
+                {
+                    var azModulePath = Path.Join(path, moduleName);
+                    DependencyManagementUtils.EmptyDirectory(azModulePath);
+                }
+
+                // Save the module to the given path
+                SaveModuleCommand(pwsh, logger, moduleName, latestVersion, path);
+            }
+        }
+
+        /// <summary>
+        /// Downloads a PowerShell module from the PSGallery to the given path.
+        /// </summary>
+        internal void SaveModuleCommand(
+            System.Management.Automation.PowerShell pwsh,
+            ILogger logger,
+            string moduleName,
+            string version,
+            string path)
+        {
+            Exception exception = null;
+
+            // If the destination path does not exist, create it.
+            if (!Directory.Exists(path))
+            {
+                Directory.CreateDirectory(path);
+            }
+
+            try
+            {
+                pwsh.AddCommand("PowerShellGet\\Save-Module")
+                    .AddParameter("Repository", Repository)
+                    .AddParameter("Name", moduleName)
+                    .AddParameter("RequiredVersion", version)
+                    .AddParameter("path", path)
+                    .AddParameter("Force", true)
+                    .AddParameter("ErrorAction", "Stop")
+                    .InvokeAndClearCommands();
+            }
+            catch (Exception e)
+            {
+                exception = e;
+                throw;
```

Comment:
I think this exception should be made an inner exception of the dependency exception that you will throw, and the message from this exception should be incorporated into the message of the outer exception.
So you don't need to catch it here.

Reply:
Thank you for this suggestion --this makes perfect sense. I have updated the code, so now we append the inner exception message (with what actual failed) to the outer exception. The error message will look like this

Fail to install function app dependencies. Error: '{0}'. The exception message will be created in the delegate that calls DependencyManager.InstallManagedDependencyModule.